src/main.cc

```diff
@@ -296,8 +318,653 @@ uint64_t rotr64 (uint64_t n, unsigned int c)
+    	}
+
+    	mmu_hit[0] = ooo_cpu[cpu].STLB.search_pml4(compare_pml4); // check for hits in the MMU cache of the PML4 level
+    	mmu_hit[1] = ooo_cpu[cpu].STLB.search_pdp(compare_pdp);   // check for hits in the MMU cache of the PDP level
+
+    	// if you use 2MB pages there is one page table level less to search on
+    	if(LOG2_PAGE_SIZE == 12)
+        	mmu_hit[2] = ooo_cpu[cpu].STLB.search_pd(compare_pd); // check for hits in the MMU cache of the PD level
+
+    	// update the MMU cache levels
+        ooo_cpu[cpu].STLB.update_pml4(ooo_cpu[cpu].STLB.mmu_timer, compare_pml4); // update the content of the MMU cache of the PML4 level if needed
+        ooo_cpu[cpu].STLB.update_pdp(ooo_cpu[cpu].STLB.mmu_timer, compare_pdp);   // update the content of the MMU cache of the PDP level if needed
+        if(LOG2_PAGE_SIZE == 12)
+        	ooo_cpu[cpu].STLB.update_pd(ooo_cpu[cpu].STLB.mmu_timer, compare_pd); // update the content of the MMU cache of the PD level if needed
+
+		uint64_t cr3 = 0x200000; // defining the CR3 register pointing address 
+        uint64_t pt_index, pd_index, pdp_index, pml4_index;
+
+		// grouping the v@ bits ... note that for 2MB pages there is one page table level lookup less
+        if(LOG2_PAGE_SIZE == 12){
+            pt_index   = (vpage       & 0x00000000001ff);
+            pd_index   = ((vpage>>9)  & 0x00000000001ff);
+            pdp_index  = ((vpage>>18) & 0x00000000001ff);
+            pml4_index = ((vpage>>27) & 0x00000000001ff);
+        }
+        else{
+            pt_index   = (vpage       & 0x00000000001ff);
+            pdp_index  = ((vpage>>9)  & 0x00000000001ff);
+            pml4_index = ((vpage>>18) & 0x00000000001ff);
+        }
+
+		uint64_t pml42s, pdp2s, pd2s, pt2s;
+
+		// calculating the address that stores the pointer to the next page table level
+        if(LOG2_PAGE_SIZE == 12){
+            pml42s = cr3 + pml4_index * 8;
+
+            pdp2s = cr3 + 512 * 8 \
+                    + pml4_index * 512 * 8 \
+                    + pdp_index * 8;
+
+            pd2s  = cr3 + 512 * 8 \
+                   	+ 512 * 512 * 8 \
+                   	+ pml4_index * 512 * 512 * 8 \
+                   	+ pdp_index * 512 * 8 \
+                   	+ pd_index * 8;
+
+            pt2s  = cr3 + 512 * 8 \
+                   	+ 512 * 512 * 8 \
+                   	+ 512 * 512 * 512 * 8 \
+                   	+ pml4_index * 512 * 512 * 512 * 8 \
+                   	+ pdp_index * 512 * 512 * 8 \
+                   	+ pd_index * 512 * 8\
+                   	+ pt_index * 8;
+        }
+        else{
+            pml42s = cr3 + pml4_index * 8;
+
+            pdp2s  = cr3 + 512 * 8 \
+                     + pml4_index * 512 * 8 \
+                     + pdp_index * 8;
+
+            pt2s   = cr3 + 512 * 8 \
+                   	 + 512 * 512 * 8 \
+                   	 + pml4_index * 512 * 512 * 8 \
+                   	 + pdp_index * 512 * 8 \
+                   	 + pt_index * 8;
+        }
```

Comment:
Maintainance here might be difficult in the future. Consider putting some kind of `#defines`, etc. 😉

Reply:
I agree. It took me a few minutes to figure out where all of the magic numbers came from.